### DISCUSSION TITLE
Github for Jira 연동을 위한 스타일 변경 논의

### DISCUSSION THREAD

**Discussion by Kirade** [8percent/styleguide#67] [General]
조만간 Jira 이슈를 통해서 진행중인 업무들을 관리하게 될 예정입니다.
Jira 의 이슈는 Github 과 연동을 통해 발행된 이슈와 Github 의 브랜치를 연결할 수 있습니다. [Github for Jira](https://marketplace.atlassian.com/apps/1219592/github-for-jira?hosting=cloud&tab=overview) 
이를 통해 발행된 이슈를 처리하기 위한 개발의 진행상태가 어떤지 Jira 이슈에서 쉽게 파악할 수 있습니다.

이것의 연동을 위해서는 Jira 이슈의 키를 Github 의 **브랜치**, **커밋**, **PR 제목**에 기입하면됩니다.
이것을 연동시키기 위한 방법들중에 어떤것을 선택하고 운영하면 좋을지 논의가 필요합니다.

제 생각에는 브랜치에 이슈 키를 명시하는 방식이 좋을것 같다고 생각합니다.

우선 커밋의 경우 브랜치에 많은 커밋들이 생길텐데 언제 어떤 커밋에 이슈 키를 등록할지에 대한 고민이 생길것 같습니다. 커밋에 이슈 키를 등록하기 좋은 경우는 특정 이슈가 현재 브랜치가 해결하고 있는 문제와 관련이 있는 경우의 레퍼런스로 등록하는 경우에 유용할 것 같습니다. 
PR 이름의 경우, 브랜치가 PR 로 발행되기 이전까지 연결되지 않아서 개발자의 작업과 이슈의 싱크가 늦어진다는 단점이 있다고 생각했습니다.

만약 브랜치에 이슈 키를 명시하게 된다면, 기존과 비슷한 형태이지만 기존의 날짜 Prefix 자리를 대체하면 어떨까 싶습니다. 그리고 기존에 사용하던 브랜치 명을 선택적으로 입력하게 두는것을 제안합니다.

AS-IS: {날짜}_{브랜치명}
TO-BE: {이슈 키}-{설명(선택)}

예시)
- BBT-102
- BBT-234-fix-some-bugs

Reference
- [Github for Jira](https://marketplace.atlassian.com/apps/1219592/github-for-jira?hosting=cloud&tab=overview) 
- [Use Github for Jira](https://support.atlassian.com/jira-cloud-administration/docs/use-the-github-for-jira-app/)

**Comment by twins1040** [comment]
저는 PR 이름이 좋을 것 같습니다. 보통 push 하면서 pr 발행하기 때문에 브랜치명에 넣는것과 pr 이름에 넣는 것이 비슷할 것 같습니다.
반대로 브랜치명에 남기지 않으면 아래와 같은 구조를 유지할 수 있을 것 같습니다.

git repo <-> github repo <-> jira

브랜치명에 jira id 가 들어간다는 것은 github 과 관계 없이 git 이 jira 와 직접 관계를 맺는 것이기 때문에
세 개의 시스템이 모두 관계를 맺는 다소 얽혀있는 구조가 됩니다.

큰 차이는 아니지만, local 에서 봤을 때의 .git 에 저장된 내용은 순수한 브랜치명과 코드 내용을 유지하는 것이 좋은 것 같습니다. 저희가 코드에 가급적 slack 이나 confluence 링크를 넣지 않는 것 처럼요.

**Reply by Kirade** [reply]
여러 기능들에 작업을 하면서 브랜치를 올려두고 PR 은 발행하지 않는 경우도 있는 것으로 알고있는데요.
예를들면 기능A 를 만드는데 A-1, A-2, A-3 브랜치를 만들어서 작업을 한 뒤, PR 하나씩 올려가며 리뷰를 받습니다. ( 한 번에 올리면 리뷰어가 불편할 수 있으니 ) 이런 경우에는 PR 에 올라가지 않은 브랜치는 Jira 이슈에서 추적을 할 수 없게 될 것 같습니다.

개인의 취향에 따라 브랜치를 활용하는 전략이 다양하다보니 쉽지 않은 문제인것 같네요.
혹시 브랜치 이름 규칙을 기존도 유지하되 이슈 키를 브랜치 Prefix 로 활용하는것을 허용하는 방식은 어떨까요?
물론 이 경우에는 기존 브랜치 이름 규칙을 쓰는 경우에는 이슈 키를 커밋 혹은 PR 제목에 표기해야합니다.

**Comment by Kirade** [comment]
데일리 스크럼에서 관련해서 간단히 이야기나눴던 내용 공유합니다.

- 이 일을하는 궁극적인 목표는 JIRA 에서 개발자가 하는일을 잘 추적하기 위함
- PR 제목에 대괄호로 표기하여 붙이는 말머리를 이슈 키로 변경하는것도 좋은 방법인것 같다.

**Comment by ParkHyeonChae** [comment]
하나의 목적을 가진 지라티켓에서 브랜치를 따내어 업무를 진행하는 과정에서 생긴 현상을 코멘트로 남깁니다.

1. BE-1, BE-2
2. BE-1-add-some-function, BE-1-fix-some-function

몇일간 조금이나마 위 두가지방식을 혼용하여 사용해보았는데요 개인적으로 느낀 생각입니다.
우선 지라를 통해 개발자의 업무를 잘 추적하고 관리하기 위함이라는 목적을 전제로 판단하였습니다.

**1번 방식**
- PR단위와 지라티켓단위가 1:1 매핑이 되는 현상이 있었습니다.
- 개발업무 진행 도중 무수히 많은 이슈와 변수들이 생기는데 (PR이 계속 생김=지라티켓이 계속 생김) 저는 이점을 하나의 지라티켓 하위태스크를 생성함으로써 해결하려 했습니다.
- 하위태스크의 티켓 번호를 PR 제목으로 작성함으로써 업무추적이 명확해지긴 하지만 너무 많은 지라티켓이 생성됨으로 저희가 목표로 하고 있는 개발자의 업무추적과 관리에 지장을 줄 수 있을 것 같다라는 생각이 들었습니다. -> 노이즈 지라티켓이 많이 생길 수 있을 것 같다.

**2번 방식**
- 노이즈가 생길법한 지라티켓을 계속 발행하는 것이 아닌 커밋을 남기는 형태처럼 분기처리하며 목적없는 지라티켓 발급을 줄일 수 있었습니다. (추가되는 브랜치는 하나의 지라티켓 설명란 또는 댓글로 기재가 충분하였습니다.)

계속 논의했던 방법, 여러가지 방법들을 시도하면서 잘 작동하는지 체크해보겠습니다

**Comment by jseop-lim** [comment]
## 전제
**Jira 이슈 없는 GitHub PR은 없다.**

**작업 배경과 정책은 GitHub PR이 아닌 곳에 반드시 문서로 남겨야 한다.**

### [참고] Jira 보충 설명
#### Jira 용어 정리
- Jira 이슈: 프로젝트에 포함된 카드 하나하나
- Jira 이슈 키: `{프로젝트 약어}-{식별번호}` 형식의 Jira 이슈에 할당된 고유한 문자열 (e.g. BE-12)
- Jira 이슈 링크: Jira 이슈에 대한 웹 페이지로 연결되는 URL
#### Jira Auto Link 기능 적용 범위
아래 항목이 Jira 이슈 키를 포함하면 Jira 이슈 링크에 대한 하이퍼링크가 자동으로 문자열에 적용되고, Jira 이슈에 각 항목에 대한 링크와 상태가 표시됨
- Git 브랜치 이름 → GitHub에 푸시하면 Jira 이슈에 브랜치가 및 PR 링크와 상태가 연동됨
- Git 커밋 제목 & 내용 → GitHub master 브랜치에 푸시되면 Jira 이슈에 커밋이 연동됨. GitHub Code의 커밋 목록에도 Jira 이슈 링크가 연동된다.
- GitHub PR 제목 & 내용 → GitHub PR을 생성하면 Jira 이슈에 PR 링크와 상태가 연동됨 (조금 느림)

## Git 브랜치 이름
서두에 Jira 이슈 키를 넣는다. Jira 이슈의 브랜치 만들기 기능이 기본으로 제공하는 `--`를 구분자로 삼는다.

Jira 이슈 키 뒤에 브랜치 작업 내용을 생략할 수 있으나, 가급적 포함하기를 권장한다.

### 예시
- `BE-12--fix-deal-list-api` (권장)
- `BE-12`

## GitHub PR 제목
PR 제목 서두에 `[{Jira 이슈 키}]`를 삽입한다.

### 이유
- 현재 `[]` 안 내용에 대한 뚜렷한 체계가 없음
- GitHub Commit 목록에 auto link 적용됨

### 예시
- `[BE-12] 상품 목록 API 오류 해결`

## GitHub PR 내용
GitHub PR 내용은 정책과 상세를 불문하고 자유롭게 적되, 리뷰하기 편하게 친절히 적는다. 필요 시 Jira 이슈 링크, Confluence 문서 링크, 사진 등을 첨부한다.

**Reply by Kirade** [reply]
`{프로젝트 약어}_{식별번호}` -> `{프로젝트 약어}-{식별번호}`

**Reply by jseop-lim** [reply]
앗 정정 감사합니다! 반영해놓도록 하겠습니다.
